crypto/encrypt.go

```diff
@@ -0,0 +1,101 @@
+package crypto
+
+import (
+	"crypto/cipher"
+	"crypto/rand"
+	"errors"
+
+	"golang.org/x/crypto/twofish"
+)
+
+const (
+	BlockSize = twofish.BlockSize
+	KeySize   = 32
+)
+
+type (
+	EncryptionKey [KeySize]byte
+)
+
+// Encryption and decryption of []bytes is supported, but requires keeping some
+// overhead information. Also, nothing is done in-place, which means the
+// functions are more memory intensive, but that's not generally the bottleneck
+// when doing encryption. Developers are less likely to make mistakes when
+// inputs and outputs are different memory, which is why this method has been
+// chosen. Additionally, the overhead of using encryption is more transparent.
+
+// GenerateEncryptionKey produces a key that can be used for encrypting and
+// decrypting files.
+func GenerateEncryptionKey() (key EncryptionKey, err error) {
+	_, err = rand.Read(key[:])
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// EncryptBytes encrypts a []byte using a key. The padded ciphertext, iv, and
+// amount of padding used are returned. `plaintext` is not overwritten.
+func EncryptBytes(key EncryptionKey, plaintext []byte) (ciphertext []byte, iv []byte, padding int, err error) {
+	// Determine the length needed for padding. The ciphertext must be padded
+	// to a multiple of BlockSize.
+	padding = BlockSize - (len(plaintext) % BlockSize)
+	if padding == BlockSize {
+		padding = 0
+	}
+
+	// Create the padded + unencrypted ciphertext.
+	ciphertext = make([]byte, len(plaintext)+padding)
+	copy(ciphertext, plaintext)
+
+	// Create the iv.
+	iv = make([]byte, BlockSize)
+	_, err = rand.Read(iv)
+	if err != nil {
+		return
+	}
+
+	// Encrypt the ciphertext.
+	blockCipher, err := twofish.NewCipher(key[:])
+	if err != nil {
+		return
+	}
+	encrypter := cipher.NewCBCEncrypter(blockCipher, iv)
+	encrypter.CryptBlocks(ciphertext, ciphertext)
+
+	return
+}
+
+// DecryptBytes decrypts a ciphertext using a key, an iv, and a volume of
+// padding. `ciphertext` is not overwritten. The plaintext is returned.
+func DecryptBytes(key EncryptionKey, ciphertext []byte, iv []byte, padding int) (plaintext []byte, err error) {
+	// Verify the iv is the correct length.
+	if len(iv) != BlockSize {
+		err = errors.New("iv is not correct size")
+		return
+	}
+	if len(ciphertext)%BlockSize != 0 {
+		err = errors.New("ciphertext is not correct size")
+		return
+	}
+	if padding > len(ciphertext) {
+		err = errors.New("stated padding is longer than the ciphertext")
+		return
+	}
+
+	// Allocate the plaintext.
+	plaintext = make([]byte, len(ciphertext))
+
+	// Decrypt the ciphertext.
+	blockCipher, err := twofish.NewCipher(key[:])
+	if err != nil {
+		return
+	}
+	decrypter := cipher.NewCBCDecrypter(blockCipher, iv)
+	decrypter.CryptBlocks(plaintext, ciphertext)
+
+	// Remove the padding.
+	plaintext = plaintext[:len(ciphertext)-padding]
```

Comment:
hmmm. We could pass around a length instead of a padding and that might clean a few things up. It's not generally how I reason about slices though, I typically think in terms of:

[thing one][thing two][thing three]. So to me having a padding makes sense because it becomes [thing three]. Otherwise you think about [thing one] and [thing two], but the slice is longer than each and you implicitly figure out that there's padding at the end which composes the extra stuff. 

I think it's less clear that there's padding if you're passing around a length that doesn't match the size of the returned data.